### PR TITLE
Convert cluster.openshift.io/Network in to the NetworkConfig CRD; update Network status

### PR DESCRIPTION
**Pull request by squeed**
This adds a second controller that watches the cluster-level network configuration. If changes are made, it merges those in to (or creates) the CRD configuration.

When the changes are applied, the "real" controller then updates the Network status object.

### CODE REVIEW

**Comment by danwinship** [comment]
But an admin can directly modify the CRD as well, right? (In fact, they have to, to configure some options). What happens if they modify the CRD to contradict the cluster config?

**Comment by squeed** [comment]
@danwinship yeah, I was thinking about that. I can see two possibilities. Either we say "hey, don't do that!", or we remove those fields from the CRD.

Removing overlapping fields from the CRD seems like the right way to me. I don't expect us to be adding much more in the way of fields to the Cluster object (except maybe Isolation).

I'll update the PR (and simplify most of the controller loop) to do this. We'll have to keep the duplicate fields for now, and roll it out in stages.

FWIW, part of this PR makes it possible to get networking up with just the cluster Network object (by defaulting to NetworkPolicy mode for openshift-sdn.

**Comment by squeed** [comment]
Okay, updated this. Now, if there are any fields that were improperly updated in the "downstream" object, revert those changes when we reconcile.

**Comment by squeed** [comment]
Odd, operator logs seem fine.
/retest

**Comment by squeed** [comment]
Hit another AWS quota issue: `Jan 07 21:30:41.565 W persistentvolume=pvc-11be8786-12c3-11e9-ab14-0a9830816660 error deleting EBS volume "vol-03b08c117479245b2": "RequestLimitExceeded: Request limit exceeded.\n\tstatus code: 503, request id: b4ada13a-d246-40a0-9f2e-c95ab2e81a1b"`

/retest

**Comment by squeed** [comment]
Flake city.
/retest

**Comment by squeed** [comment]
All green. @danwinship, would you mind reviewing?

File: pkg/network/cluster_config.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	// - they do not overlap and do not overlap with the service cidr
+	for _, cnet := range clusterConf.ClusterNetwork {
+		_, cidr, err := net.ParseCIDR(cnet.CIDR)
+		if err != nil {
+			return errors.Errorf("could not parse spec.clusterNetwork %s", cnet.CIDR)
+		}
+		size, _ := cidr.Mask.Size()
+		// The comparison is inverted; smaller number is larger block
+		if cnet.HostPrefix < uint32(size) {
+			return errors.Errorf("hostPrefix %d is larger than its cidr %s",
```
Comment (danwinship) [inline pkg/network/cluster_config.go]: To match origin's validation, this should be `<=` not `<`. Also it checks that `HostSubnetLength >= 2` aka `HostPrefix <= 30`

File: pkg/network/cluster_config.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// StatusFromConfig generates the cluster NetworkStatus from the currently applied configuration
+func StatusFromConfig(operConf *netopv1.NetworkConfigSpec) configv1.NetworkStatus {
+	status := configv1.NetworkStatus{
+		ServiceNetwork: []string{operConf.ServiceNetwork},
+		NetworkType:    string(operConf.DefaultNetwork.Type),
+	}
+
+	// Flip flip flip the bits
```
Comment (danwinship) [inline pkg/network/cluster_config.go]: should we fix the operator config to match the cluster config?

File: pkg/controller/networkconfig/cluster.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	// Merge the cluster config in to the downstream CRD
+	network.MergeClusterConfig(&cfg.Spec, clusterConf.Spec)
+	if reflect.DeepEqual(cfg.Spec, existing.Spec) {
+		return nil
+	}
+
+	// If there are changes to the "downstream" networkconfig, commit it back
+	// to the apiserver
+	log.Println("WARNING: NetworkConfig.networkoperator.openshift.io has fields being overwritten by Network.cluster.openshift.io configuration")
```
Comment (danwinship) [inline pkg/controller/networkconfig/cluster.go]: name of the second api object is wrong; should be `Network.config.openshift.io`

File: pkg/controller/networkconfig/cluster.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	cfg.TypeMeta = metav1.TypeMeta{APIVersion: "networkoperator.openshift.io/v1", Kind: "NetworkConfig"}
```
Comment (danwinship) [inline pkg/controller/networkconfig/cluster.go]: why is this needed?

File: pkg/controller/networkconfig/cluster.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "failed to transmute operator config")
+	}
+	if err = apply.ApplyObject(context.TODO(), r.client, us); err != nil {
+		return errors.Wrapf(err, "could not apply (%s) %s/%s", cfg.GroupVersionKind(), cfg.GetNamespace(), cfg.GetName())
+	}
+	return nil
+}
+
+// ClusterNetworkStatus generates the cluster.openshift.io/v1/Network.Status
```
Comment (danwinship) [inline pkg/controller/networkconfig/cluster.go]: **config**.openshift.io

File: pkg/controller/networkconfig/networkconfig_controller.go [hunk 6]
```diff
@@ -163,6 +167,22 @@ func (r *ReconcileNetworkConfig) Reconcile(request reconcile.Request) (reconcile
...
 
+	// Update cluster.openshift.io/Network.Status
+	status, err := r.ClusterNetworkStatus(context.TODO(), instance)
+	if err != nil {
+		err = errors.Wrapf(err, "could not generate network status")
+		log.Println(err)
+		return reconcile.Result{}, err
+	}
+	if status != nil {
+		// Don't set the owner reference
```
Comment (danwinship) [inline pkg/controller/networkconfig/networkconfig_controller.go]: i don't know what this comment means

File: pkg/util/ip/addr.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+type IPPool struct {
+	cidrs []net.IPNet
+}
+
+func (p *IPPool) Add(cidr net.IPNet) error {
+	for _, n := range p.cidrs {
+		if netsOverlap(n, cidr) {
+			return errors.Errorf("cidrs %s and %s overlap",
```
Comment (danwinship) [inline pkg/util/ip/addr.go]: CIDRs

File: pkg/controller/clusterconfig/clusterconfig_controller.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch the NetworkConfig instance
+	instance := &configv1.Network{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
```
[the remark this answers is not in the document]
Comment (squeed) [reply pkg/controller/clusterconfig/clusterconfig_controller.go]: This comment is boilerplate automatically generated by the operator-sdk, FWIW. Since we set the ownerReference, the usual kubernetes garbage-collection rules apply.

However, we don't set the owner reference in this case, because I decided that we don't want to delete the CRD. This is only because the network is special, and deleting the network configuration is likely to be a serious error. In most cases, this is not expected behavior, and I could be convinced to change this.

I'll delete the misleading comment, though.

File: pkg/controller/networkconfig/cluster.go (hunk 4 above)
Comment (danwinship) [reply pkg/controller/networkconfig/cluster.go]: maybe this is one of those things where you're mutating an object you got from a cache rather than DeepCopying it and mutating the copy?

File: pkg/network/cluster_config.go (hunk 1 above)
Comment (squeed) [reply pkg/network/cluster_config.go]: By my reading, it seems to allow a HostPrefix that is the same size as the CIDR? I'm looking [here](https://github.com/openshift/origin/blob/master/pkg/network/apis/network/validation/validation.go#L94)

**Comment by squeed** [comment]
@danwinship Thanks for the excellent review. suggestions incorporated, PTAL.

File: pkg/controller/networkconfig/cluster.go (hunk 4 above)
Comment (squeed) [reply pkg/controller/networkconfig/cluster.go]: Probably not that, but something similar. The client always copies in to a receiving object, and all the types should be registered with the schema. Just confusing.

File: pkg/network/cluster_config.go (hunk 1 above)
Comment (danwinship) [reply pkg/network/cluster_config.go]: huh... somehow i looked at that same code and got the wrong answer.

**Comment by squeed** [comment]
@danwinship can I get a LGTM?

File: README.md [hunk 9]
```diff
@@ -4,15 +4,44 @@ The Cluster Network Operator installs and upgrades the networking components on
...
-The network operator has a complex configuration, but most parameters have a sensible default.
+The network operator gets its configuration from two objects: the Cluster and the Operator configuration. Most users only need to create the Cluster configuration - the operator will generate its configuration automatically. If you need finer-grained configuration of your network, you will need to create both configurations.
+
+Any changes to the Cluster configuration are propagated down in to the Operator configuration.
+
+
+#### Configuration objects
+*Cluster config*
+- *Type Name*: `Network.config.openshift.io`
+- *Instance Name*: `cluster`
```
Comment (danwinship) [inline README.md]: (I assume this name matches what other people are doing?)

File: README.md [hunk 10]
```diff
@@ -4,15 +4,44 @@ The Cluster Network Operator installs and upgrades the networking components on
...
+- *View Command*: `oc get Network.openshift.io cluster -oyaml`
```
Comment (danwinship) [inline README.md]: Network.**config**.openshift.io

File: README.md [hunk 11]
```diff
@@ -4,15 +4,44 @@ The Cluster Network Operator installs and upgrades the networking components on
...
+  name: cluster
+spec:
+  serviceNetwork: ["172.30.0.0/16"]
+  clusterNetwork: 
+    - cidr: "10.128.0.0/14"
+      hostPrefix: 23 
+  networkType: OpenShiftSDN
+```
+
+*Operator Config*
```
Comment (danwinship) [inline README.md]: (Clarify that the operator config in this example would have been auto-generated. Or alternatively, add a non-auto-generatable line to it and then explain that everything except that one line was auto-generated.)

File: README.md [hunk 12]
```diff
@@ -4,15 +4,44 @@ The Cluster Network Operator installs and upgrades the networking components on
...
 
+Most users will be able to use the top-level OpenShift Config API, which has a [Network type](https://github.com/openshift/api/blob/master/config/v1/types_network.go#L26). The operator will automatically translate the `Network.config.openshift.io` object in to a `NetworkConfig.networkoperator.openshift.io`.
+
 When the controller has reconciled and all its dependent resources have converged, the cluster should have an installed SDN plugin and a working service network. In OpenShift, the Cluster Network Operator runs very early in the install process -- while the boostrap API server is still running.
 
 # Configuring
-The network operator has a complex configuration, but most parameters have a sensible default.
+The network operator gets its configuration from two objects: the Cluster and the Operator configuration. Most users only need to create the Cluster configuration - the operator will generate its configuration automatically. If you need finer-grained configuration of your network, you will need to create both configurations.
+
+Any changes to the Cluster configuration are propagated down in to the Operator configuration.
```
Comment (danwinship) [inline README.md]: State explicitly that in case of conflicts, the operator config will be updated to match the cluster config

File: README.md [hunk 13]
```diff
@@ -30,9 +59,14 @@ spec:
 ## Configuring IP address pools
 Users must supply at least two address pools - one for pods, and one for services. These are the ClusterNetworks and ServiceNetwork parameter. Some network plugins, such as OpenShiftSDN, support multiple ClusterNetworks. All address blocks must be non-overlapping. You should select address pools large enough to fit your anticipated workload.
 
+For future expansion, multiple `serviceNetwork` entries are allowed by the configuration but not actually supported by any network plugins. Supplying multiple addresses is invalid.
+
+Each `clusterNetwork` entry has an additional required parameter, `hostPrefix`, that specifies the address size to assign to assign to each individual node.  For example
```
[the remark this answers is not in the document]
Comment (danwinship) [reply README.md]: so either add the example or remove "For example"

File: README.md [hunk 14]
```diff
@@ -30,9 +59,14 @@ spec:
...
+
+IP address pulls are always read from the Cluster configuration and propagated "downwards" in to the Operator configuration. Any changes to the Operator configuration will be ignored.
```
[the remark this answers is not in the document]
Comment (danwinship) [reply README.md]: fix that too

File: README.md [hunk 15]
```diff
@@ -4,15 +4,44 @@ The Cluster Network Operator installs and upgrades the networking components on
...
+#### Example configurations
 
+*Cluster Config*
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  serviceNetwork: ["172.30.0.0/16"]
```
Comment (danwinship) [inline README.md]: maybe use the expanded syntax for parallel-ness with `clusterNetwork`? (And to match how `oc get` will display it.)

    serviceNetwork:
      - "172.30.0.0/16"
    clusterNetwork:
      - cidr: "10.128.0.0/14"
        hostPrefix: 23

File: pkg/controller/clusterconfig/clusterconfig_controller.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile propagates changes from the cluster config to the operator config.
+// In other words, it watches Network.config.openshift.io/v1/cluster and updates
+// NetworkConfig.networkoperator.openshift.io/v1/default.
+func (r *ReconcileClusterConfig) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling Network %s/%s\n", request.Namespace, request.Name)
```
Comment (danwinship) [inline pkg/controller/clusterconfig/clusterconfig_controller.go]: "Network" here is ambiguous. Either `Network.config.openshift.io` or `cluster config`.

Also, it's non-namespaced so don't print the namespace. (That may apply to ReconcileNetworkConfig too?)

Comment (squeed) [reply pkg/controller/clusterconfig/clusterconfig_controller.go]: Good point; fixed (for both)

File: README.md (hunk 15 above)
Comment (squeed) [reply README.md]: There, updated both to match `oc get` exactly (sans metadata).

File: README.md (hunk 9 above)
Comment (squeed) [reply README.md]: Yup. Once the installer switches from `NetworkConfig` to `Network`, we can align the names if we like.

**Comment by squeed** [comment]
@danwinship doc nits and log lines fixed, thanks.
